### ISSUE TITLE
Reading consecutive HTTPS responses with empty bodies hangs

### ISSUE THREAD

**Issue by philsttr** [spring-projects/spring-framework#22819]
If I submit two HTTPS requests using WebClient (using reactor-netty), and the second response has an empty body (content-length: 0), then the client hangs and does not emit a response.
When I do the same over HTTP, it works.  So, this appears isolated to HTTPS.

| module | version |
|----|----|
| spring-webflux | 5.1.6.RELEASE |
| reactor-core | 3.2.8.RELEASE|
| reactor-netty | 0.8.6.RELEASE|
| netty |4.1.34.Final |

I initially discovered this when running against a real server that returned an empty protobuf payload.

I have a created a unit test ([project attached](https://github.com/spring-projects/spring-framework/files/3096139/webclient_empty_body.zip)) that executes requests against an okhttp3 MockWebServer to demonstrate the issue.
The unit test has two methods that each execute two requests against the mock server, with the only difference being HTTP vs HTTPS.
* The HTTP test passes.
* The HTTPS test fails with a timeout exception waiting on the second response.

Here is what I have noticed in trace logs.

For the second HTTP request (which succeeds):

```
2019-04-18 13:41:20,635 INFO  [main]: REQUEST #2
2019-04-18 13:41:20,635 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] onStateChange(POST{uri=/, connection=PooledConnection{channel=[id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670]}}, [disconnecting])
2019-04-18 13:41:20,635 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Releasing channel
2019-04-18 13:41:20,635 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Channel cleaned, now 0 active connections and 1 inactive connections
2019-04-18 13:41:20,635 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] READ COMPLETE
2019-04-18 13:41:20,636 TRACE [main]: [47c64cfe] HTTP POST http://127.0.0.1:60670/, headers={}
2019-04-18 13:41:20,637 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Channel acquired, now 1 active connections and 0 inactive connections
2019-04-18 13:41:20,637 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Handler is being applied: {uri=http://127.0.0.1:60670/, method=POST}
2019-04-18 13:41:20,638 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] WRITE: 154B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 50 4f 53 54 20 2f 20 48 54 54 50 2f 31 2e 31 0d |POST / HTTP/1.1.|
|00000010| 0a 75 73 65 72 2d 61 67 65 6e 74 3a 20 52 65 61 |.user-agent: Rea|
|00000020| 63 74 6f 72 4e 65 74 74 79 2f 30 2e 38 2e 36 2e |ctorNetty/0.8.6.|
|00000030| 52 45 4c 45 41 53 45 0d 0a 68 6f 73 74 3a 20 31 |RELEASE..host: 1|
|00000040| 32 37 2e 30 2e 30 2e 31 3a 36 30 36 37 30 0d 0a |27.0.0.1:60670..|
|00000050| 61 63 63 65 70 74 3a 20 2a 2f 2a 0d 0a 43 6f 6e |accept: */*..Con|
|00000060| 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 78 74 2f |tent-Type: text/|
|00000070| 70 6c 61 69 6e 3b 63 68 61 72 73 65 74 3d 55 54 |plain;charset=UT|
|00000080| 46 2d 38 0d 0a 63 6f 6e 74 65 6e 74 2d 6c 65 6e |F-8..content-len|
|00000090| 67 74 68 3a 20 30 0d 0a 0d 0a                   |gth: 0....      |
+--------+-------------------------------------------------+----------------+
2019-04-18 13:41:20,638 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] FLUSH
2019-04-18 13:41:20,639 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] WRITE: 0B
2019-04-18 13:41:20,639 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] FLUSH
2019-04-18 13:41:20,639 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] onStateChange(POST{uri=/, connection=PooledConnection{channel=[id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670]}}, [request_sent])
2019-04-18 13:41:20,640 INFO  [MockWebServer /127.0.0.1:60671]: MockWebServer[60670] received request: POST / HTTP/1.1 and responded: HTTP/1.1 200 OK
2019-04-18 13:41:20,640 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] READ: 38B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0d |HTTP/1.1 200 OK.|
|00000010| 0a 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a |.Content-Length:|
|00000020| 20 30 0d 0a 0d 0a                               | 0....          |
+--------+-------------------------------------------------+----------------+
2019-04-18 13:41:20,640 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Received response (auto-read:false) : [Content-Length=0]
2019-04-18 13:41:20,641 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] onStateChange(POST{uri=/, connection=PooledConnection{channel=[id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670]}}, [response_received])
2019-04-18 13:41:20,641 TRACE [reactor-http-nio-9]: [47c64cfe] Response 200 OK, headers={masked}
2019-04-18 13:41:20,641 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] Received last HTTP packet

```

For the second HTTPS request (which hangs):

```
2019-04-18 13:41:15,588 INFO  [main]: REQUEST #2
2019-04-18 13:41:15,588 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] Channel cleaned, now 0 active connections and 1 inactive connections
2019-04-18 13:41:15,588 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] READ COMPLETE
2019-04-18 13:41:15,589 TRACE [main]: [3e134896] HTTP POST https://127.0.0.1:60640/, headers={}
2019-04-18 13:41:15,591 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] Channel acquired, now 1 active connections and 0 inactive connections
2019-04-18 13:41:15,591 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] Handler is being applied: {uri=https://127.0.0.1:60640/, method=POST}
2019-04-18 13:41:15,592 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] WRITE: 154B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 50 4f 53 54 20 2f 20 48 54 54 50 2f 31 2e 31 0d |POST / HTTP/1.1.|
|00000010| 0a 75 73 65 72 2d 61 67 65 6e 74 3a 20 52 65 61 |.user-agent: Rea|
|00000020| 63 74 6f 72 4e 65 74 74 79 2f 30 2e 38 2e 36 2e |ctorNetty/0.8.6.|
|00000030| 52 45 4c 45 41 53 45 0d 0a 68 6f 73 74 3a 20 31 |RELEASE..host: 1|
|00000040| 32 37 2e 30 2e 30 2e 31 3a 36 30 36 34 30 0d 0a |27.0.0.1:60640..|
|00000050| 61 63 63 65 70 74 3a 20 2a 2f 2a 0d 0a 43 6f 6e |accept: */*..Con|
|00000060| 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 78 74 2f |tent-Type: text/|
|00000070| 70 6c 61 69 6e 3b 63 68 61 72 73 65 74 3d 55 54 |plain;charset=UT|
|00000080| 46 2d 38 0d 0a 63 6f 6e 74 65 6e 74 2d 6c 65 6e |F-8..content-len|
|00000090| 67 74 68 3a 20 30 0d 0a 0d 0a                   |gth: 0....      |
+--------+-------------------------------------------------+----------------+
2019-04-18 13:41:15,593 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] FLUSH
2019-04-18 13:41:15,593 DEBUG [reactor-http-nio-4]: [id: 0x518b335c, L:/127.0.0.1:60666 - R:/127.0.0.1:60640] WRITE: 0B
2019-04-18 13:41:15,594 INFO  [MockWebServer /127.0.0.1:60666]: MockWebServer[60640] received request: POST / HTTP/1.1 and responded: HTTP/1.1 200 OK
<<< HANG >>>
```

At the point of hanging, it seems the second HTTPS request does _not_ log the following (which is seen in HTTP), indicating that maybe it thinks the request has not been fully sent yet, and therefore does not start waiting on the response...

```
2019-04-18 13:41:20,639 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] FLUSH
2019-04-18 13:41:20,639 DEBUG [reactor-http-nio-9]: [id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670] onStateChange(POST{uri=/, connection=PooledConnection{channel=[id: 0xa50ef88e, L:/127.0.0.1:60671 - R:/127.0.0.1:60670]}}, [request_sent])
```

I'm not sure if this is a reactor-netty issue, or a WebClient issue.  I'll start here, but I'll be happy to file an issue on reactor-netty if needed.

**Comment by philsttr** [comment]
After experimenting a bit more, it looks like this is related to posting an empty _request_ body.

If I change the request body sent in the unit test from this:
```java
		return webClient.post()
				.syncBody("")
```
to this...
```java
		return webClient.post()
				.syncBody("hello")
```

Then both test cases work.

It continues to work if I change the response to non-empty body (while adding `.expectNextCount(1)` to the `StepVerifier`).

To recap:
* Always works over HTTP (regardless of request/response body length)
* Always works over HTTPS if request body length > 0 (regardless of response body length)
* The first request always works.
* The _second_ request hangs when using an empty _request_ body over HTTPS

**Comment by philsttr** [comment]
I believe this is a reactor-netty issue.  I can reproduce using `HttpClient` without `WebClient`

Closing this in favor of https://github.com/reactor/reactor-netty/issues/719
